File: Go/ExampleTests/common/shippinginput_string.go
package common

type ShippingInputString struct {
	TotalPrice string
	ShippingCost string
	Notes string
}

func NewShippingInputStringFromSlice(v []string) ShippingInputString {
	s := ShippingInputString{}
	if len(v) > 0 { s.TotalPrice = v[0] }
	if len(v) > 1 { s.ShippingCost = v[1] }
	if len(v) > 2 { s.Notes = v[2] }
	return s
}

// NewShippingInputStringFromText builds from the text form, e.g. Money as "25 USD".
func NewShippingInputStringFromText(text string) ShippingInputString {
	parts := RequireTokens(text, 3, "ShippingInput")
	return ShippingInputString{
		TotalPrice: parts[0],
		ShippingCost: parts[1],
		Notes: parts[2],
	}
}

func (s ShippingInputString) String() string {
	return Token(s.TotalPrice) + " " + Token(s.ShippingCost) + " " + Token(s.Notes)
}

func (s ShippingInputString) Equals(o ShippingInputString) bool {
	return DNCEqual(s.TotalPrice, o.TotalPrice) &&
		DNCEqual(s.ShippingCost, o.ShippingCost) &&
		DNCEqual(s.Notes, o.Notes)
}

func EqualShippingInputStringSlices(a, b []ShippingInputString) bool {
	if len(a) != len(b) { return false }
	for i := range a {
		if !a[i].Equals(b[i]) { return false }
	}
	return true
}

File: Rust/ExampleTests/common/shippinginput_string.rs
#![allow(dead_code, unused_imports, unused_variables)]

#[derive(Debug, Clone, Default)]
pub struct ShippingInputString {
    pub total_price: String,
    pub shipping_cost: String,
    pub notes: String,
}

impl ShippingInputString {
    pub fn from_vec(v: &[&str]) -> Self {
        Self {
            total_price: v.get(0).copied().unwrap_or("").to_string(),
            shipping_cost: v.get(1).copied().unwrap_or("").to_string(),
            notes: v.get(2).copied().unwrap_or("").to_string(),
        }
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    pub fn from_text(text: &str) -> Self {
        let parts = crate::common::tokens::require(text, 3, "ShippingInput");
        Self {
            total_price: parts[0].clone(),
            shipping_cost: parts[1].clone(),
            notes: parts[2].clone(),
        }
    }
}

impl std::fmt::Display for ShippingInputString {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", [crate::common::tokens::token(&self.total_price), crate::common::tokens::token(&self.shipping_cost), crate::common::tokens::token(&self.notes)].join(" "))
    }
}

impl PartialEq for ShippingInputString {
    fn eq(&self, other: &Self) -> bool {
        crate::common::dnc_equal(&self.total_price, &other.total_price)
            && crate::common::dnc_equal(&self.shipping_cost, &other.shipping_cost)
            && crate::common::dnc_equal(&self.notes, &other.notes)
    }
}
